python/ray/services.py

```diff
@@ -505,62 +442,106 @@ def start_redis(node_ip_address,
+    if not use_credis:
+        assigned_port, _ = _start_redis_instance(
+            node_ip_address=node_ip_address,
+            port=port,
+            redis_max_clients=redis_max_clients,
+            stdout_file=redis_stdout_file,
+            stderr_file=redis_stderr_file,
+            cleanup=cleanup)
+    else:
+        assigned_port, _ = _start_redis_instance(
+            node_ip_address=node_ip_address,
+            port=port,
+            redis_max_clients=redis_max_clients,
+            stdout_file=redis_stdout_file,
+            stderr_file=redis_stderr_file,
+            cleanup=cleanup,
+            executable=CREDIS_EXECUTABLE,
+            # It is important to load the credis module BEFORE the ray module,
+            # as the latter contains an extern declaration that the former
+            # supplies.
+            modules=[CREDIS_MASTER_MODULE, REDIS_MODULE])
     if port is not None:
         assert assigned_port == port
     port = assigned_port
     redis_address = address(node_ip_address, port)
 
     # Register the number of Redis shards in the primary shard, so that clients
     # know how many redis shards to expect under RedisShards.
-    redis_client = redis.StrictRedis(host=node_ip_address, port=port)
-    redis_client.set("NumRedisShards", str(num_redis_shards))
+    primary_redis_client = redis.StrictRedis(host=node_ip_address, port=port)
+    primary_redis_client.set("NumRedisShards", str(num_redis_shards))
 
     # Put the redirect_worker_output bool in the Redis shard so that workers
     # can access it and know whether or not to redirect their output.
-    redis_client.set("RedirectOutput", 1 if redirect_worker_output else 0)
+    primary_redis_client.set("RedirectOutput", 1
+                             if redirect_worker_output else 0)
 
     # Store version information in the primary Redis shard.
-    _put_version_info_in_redis(redis_client)
+    _put_version_info_in_redis(primary_redis_client)
 
     # Start other Redis shards. Each Redis shard logs to a separate file,
     # prefixed by "redis-<shard number>".
     redis_shards = []
     for i in range(num_redis_shards):
         redis_stdout_file, redis_stderr_file = new_log_files(
             "redis-{}".format(i), redirect_output)
-        redis_shard_port, _ = start_redis_instance(
-            node_ip_address=node_ip_address,
-            port=redis_shard_ports[i],
-            redis_max_clients=redis_max_clients,
-            stdout_file=redis_stdout_file,
-            stderr_file=redis_stderr_file,
-            cleanup=cleanup)
+        if not use_credis:
+            redis_shard_port, _ = _start_redis_instance(
+                node_ip_address=node_ip_address,
+                port=redis_shard_ports[i],
+                redis_max_clients=redis_max_clients,
+                stdout_file=redis_stdout_file,
+                stderr_file=redis_stderr_file,
+                cleanup=cleanup)
+        else:
+            assert num_redis_shards == 1, \
```

Comment:
How difficult will it be to remove this restriction? (in a subsequent PR)

Reply:
To make this work we need both of the following
* Vanilla support for sharding for RAY_USE_NEW_GCS=on (related to new GCS API; not related to chain)
* (related to chain) code changes to the master, to make it support handling multiple chains; currently the master module assumes 1 master/chain.

Both items need non-negligible amount of engineering work, I think.